spec/requests/query_controller_spec.rb

```diff
@@ -170,6 +170,94 @@ def run_query(id, params = {})
         expect(response_json["errors"].first).to match(/ValidationError/)
       end
 
+      context "when rate limited" do
+        def unlimited_request(query_id, headers = {})
+          post "/admin/plugins/explorer/queries/#{query_id}/run.json",
+               params: {
+                 params: {}.to_json,
+               },
+               headers: headers
+
+          expect(response.status).to eq(200)
+        end
+
+        def limited_request(query_id, headers = {})
+          post "/admin/plugins/explorer/queries/#{query_id}/run.json",
+               params: {
+                 params: {}.to_json,
+               },
+               headers: headers
+
+          expect(response.status).to eq(429)
+          expect(response.parsed_body["extras"]).to eq(
+            { "wait_seconds" => 9, "time_left" => "9 seconds" },
+          )
+        end
+
+        it "limits query runs from API when using block mode" do
+          GlobalSetting.stubs(:max_data_explorer_api_reqs_per_10_seconds).returns(1)
```

Comment:
We have a helper for this stubbing, see:

https://github.com/discourse/discourse/blob/3a9a534881135ac8928266b1d02bc679f2b49b31/spec/lib/global_path_spec.rb#L30-L44

https://github.com/discourse/discourse/blob/3a9a534881135ac8928266b1d02bc679f2b49b31/spec/rails_helper.rb#L478-L484